design-proposals/vhostuser/vhostuser-interface.md

```diff
@@ -0,0 +1,166 @@
+            vhostuser: {}
+          useVirtioTransitional: true
+      networks:
+      - name: vhost-user-vn-blue
+        multus:
+          networkName: vn-blue
+```
+No additional definition of vloumes or annotations are required. The DownwardAPI will provide all the necessary annotations required for vhostUser Interface.
+
+## Scalability
+- There should not be any scalability issues, as the feature is to create an interface when useful.
+- VM migration works the same way, as the only change is having a new interace, which only requires a dataplane of the host in userpsace.
+
+## Update/Rollback Compatibility
+- Should have no impact.
+
+## Functional Testing Approach
+Functional test can:
+
+- Create VM with vhostUser interface
+- Create VM with multiple interface types. (Bridge + vhostuser type)
+
+# Implementation Phases
+- The design will be implemented in two phases:
+    - Add DPDK support on kubevirt/kubevirtci
+        - Enable OVS with DPDK
+        - Enable an option in gocli for DPDK support
+        - Add UserSpace CNI in the setup instead of or along with calico.
+    - Add vhostUser Interface type in kubevirt/kubevirt
+        - Add vHostUser Interface type
+        - Create Pod template to support the Interface
+        - Create appropriate virsh xml elements
+        - Add E2E test to send traffic between the 2 VMs created.
+
+## Annex
+
+
+A sample DownwardAPi mght look this way:
+
+```yaml
+k8s.v1.cni.cncf.io/network-status="[{\n    \"name\": \"k8s-kubemanager-kubernetes-CNI/default-podnetwork\",\n    \"interface\": \"eth0\",\n    \"ips\": [\n        \"10.244.2.4\"\n    ],\n    \"mac\": \"02:d7:b3:27:b3:b2\",\n    \"default\": true,\n    \"dns\": {}\n},{\n    \"name\": \"kubevirttest/vn-blue\",\n    \"interface\": \"net1\",\n    \"ips\": [\n        \"19.1.1.2\"\n    ],\n    \"mac\": \"02:c6:da:d3:ab:07\",\n    \"dns\": {},\n    \"device-info\": {\n        \"type\": \"vhost-user\",\n        \"vhost-user\": {\n            \"mode\": \"server\",\n            \"path\": \"2ea1931c-2935-net1\"\n        }\n    }\n}]"
+k8s.v1.cni.cncf.io/networks="[{\"interface\":\"net1\",\"name\":\"vn-blue\",\"namespace\":\"kubevirttest\",\"cni-args\":{\"interface-type\":\"virtio\"}}]"
+kubectl.kubernetes.io/default-container="compute"
+kubernetes.io/config.seen="2023-05-16T16:50:04.224937236Z"
+kubernetes.io/config.source="api"
+kubevirt.io/domain="vm-single-virtio"
+kubevirt.io/migrationTransportUnix="true"
+post.hook.backup.velero.io/command="[\"/usr/bin/virt-freezer\", \"--unfreeze\", \"--name\", \"vm-single-virtio\", \"--namespace\", \"kubevirttest\"]"
+post.hook.backup.velero.io/container="compute"
+pre.hook.backup.velero.io/command="[\"/usr/bin/virt-freezer\", \"--freeze\", \"--name\", \"vm-single-virtio\", \"--namespace\", \"kubevirttest\"]"
+pre.hook.backup.velero.io/container="compute"
+```
+The CNI should update the /ect/podnetinfo/annotaions with something similar, to update the annotations of the virt-launcher pod.
```

Comment:
Should the path be /etc/podnetinfo/annotations?

Reply:
Yes, the path is set that way to update the pod annotations.

Reply:
Okay. There is a typo in the path on line 116.